### ISSUE TITLE
MSH4.1

### ISSUE THREAD

**Issue by gdmcbain** [nschloe/pygmsh#240]
The latest wave of [gmsh#444 MSH 4.x format revisions](https://gitlab.onelab.info/gmsh/gmsh/issues/444) has broken pygmsh.generate_mesh; e.g. in the test suite, '42 failed, 4 passed' and for a minimal example, try [segment.py.gz](https://github.com/nschloe/pygmsh/files/2903635/segment.py.gz)

```python
geom = Geometry()
geom.add_line(*(geom.add_point([x, 0., 0.]) for x in [0., 1.]))
print(generate_mesh(geom, dim=1))
```
which raises an _Erreur de segmentation_.  

This is with Gmsh 4.2.0-git-d474ec1e2.

**Comment by gdmcbain** [comment]
Appending
```python
mesh_file_type='mesh'
```
to the arguments of `generate_mesh` works in the minimal example, segment.py.

**Comment by gdmcbain** [comment]
But `mesh_file_type='msh2'` raises
> AssertionError: Could not deduce file format from extension '.msh2'.

which is a different error.  I'll try to fix that first as it will provide a workaround.

**Comment by gdmcbain** [comment]
The GEO and MSH 4.1 files corresponding to the above minimal example are:
* [segment.geo.gz](https://github.com/nschloe/pygmsh/files/2903736/segment.geo.gz)
* [segment.msh.gz](https://github.com/nschloe/pygmsh/files/2903726/segment.msh.gz)

**Comment by nschloe** [comment]
Thanks Geordie for taking this up. At the moment, pygmsh isn't even compatible with gmsh 4.0*, but I'll take a look at this in the afternoon. I'll leave the format changes to you. :)

> _Erreur de segmentation_

Oh my.

**Comment by gdmcbain** [comment]
> At the moment, pygmsh isn't even compatible with gmsh 4.0*

Actually I've been using daily git-pulled pygmsh and gmsh, and therefore MSH4, for a while now since nschloe/meshio#304, and it's been working O. K.

It's not feature-complete—in particular there's compound entities #112 #214—but (until this issue arose this week), it's been good enough to use in several of the examples in scikit-fem and some bread-and-butter FreeFem++ work.

**Comment by gdmcbain** [comment]
This is nearly fixed with the merger of nschloe/meshio#331 ('update msh4 reader to MSH4.1') though maybe the version needs incrementing; the latest pip-updated meshio gives 45 failing tests but git-pulled master only 5.  All scikit-fem examples relying on pygmsh now work again.

This is with gmsh 4.2.3-git-89a21c30a.

The five remaining failures might be new, something to do periodicity relations that I haven't seen before.

**Comment by nschloe** [comment]
> maybe the version needs incrementing

That'll happen soon.

> The five remaining failures might be new,

Let's assume that, and open a new bug if something unexpected crops up.

Great work, @gdmcbain!
